File: v4/rust/iprange-livedb/benches/update_ipsets/scenarios.rs
use std::hint::black_box;
use std::path::{Path, PathBuf};

use iprange_livedb::{
    snapshot_to,
    validation::{validate, ValidationBudget, ValidationMode, ValidationSinkControl},
    CancellationToken, CloseOutcome, CommitDurability, CommitResult, Ipv4Key, LiveReader,
    LiveWriter, SnapshotPublicationPolicy, SnapshotSourceMode,
};

use crate::measure::{self, FileSize, Measurement};
use crate::model::{snapshot_budget, TestDatabase};
use crate::source::FeedShape;

#[path = "scenarios/direct.rs"]
mod direct;
#[path = "scenarios/membership.rs"]
mod membership;
#[path = "scenarios/read.rs"]
mod read;
#[path = "scenarios/sdk.rs"]
mod sdk;

#[derive(Clone, Debug)]
pub(crate) struct ScenarioResult {
    pub(crate) name: &'static str,
    pub(crate) size: usize,
    pub(crate) auxiliary: usize,
    pub(crate) work_units: u64,
    pub(crate) emitted_units: u64,
    pub(crate) range_records: u64,
    pub(crate) feeds: u64,
    pub(crate) measurement: Measurement,
    pub(crate) file: FileSize,
    pub(crate) private_artifacts: u64,
}

pub(crate) struct ImmutableResultSpec {
    pub(crate) name: &'static str,
    pub(crate) size: usize,
    pub(crate) auxiliary: usize,
    pub(crate) work_units: u64,
    pub(crate) emitted_units: u64,
}

pub(crate) fn run(name: &str, size: usize, auxiliary: usize) -> Result<ScenarioResult, String> {
    if size == 0 {
        return Err("size must be positive".to_owned());
    }
    match name {
        "direct-replace" => direct::replace(size),
        "nested-overwrite" => direct::nested(size),
        "first-seen-refresh" => direct::first_seen(size),
        "last-seen-refresh" => direct::last_seen(size),
        "feed-replace" => membership::replace_feed(size, auxiliary),
        "feed-first-ascending" => membership::shaped_feed(
            "feed-first-ascending",
            size,
            FeedShape::AscendingDisjoint,
            false,
        ),
        "feed-second-ascending" => membership::shaped_feed(
            "feed-second-ascending",
            size,
            FeedShape::AscendingDisjoint,
            true,
        ),
        "feed-first-descending" => membership::shaped_feed(
            "feed-first-descending",
            size,
            FeedShape::DescendingDisjoint,
            false,
        ),
        "feed-second-descending" => membership::shaped_feed(
            "feed-second-descending",
            size,
            FeedShape::DescendingDisjoint,
            true,
        ),
        "feed-first-random" => {
            membership::shaped_feed("feed-first-random", size, FeedShape::RandomDisjoint, false)
        }
        "feed-second-random" => {
            membership::shaped_feed("feed-second-random", size, FeedShape::RandomDisjoint, true)
        }
        "feed-first-overlap" => membership::shaped_feed(
            "feed-first-overlap",
            size,
            FeedShape::RandomOverlapChain,
            false,
        ),
        "feed-second-overlap" => membership::shaped_feed(
            "feed-second-overlap",
            size,
            FeedShape::RandomOverlapChain,
            true,
        ),
        "membership-import" => membership::import(size, auxiliary),
        "live-membership-lookup" => membership::live_lookup(size, auxiliary),
        "immutable-membership-lookup" => membership::immutable_lookup(size, auxiliary),
        "live-feed-scan" => membership::live_scan(size, auxiliary),
        "immutable-feed-scan" => membership::immutable_scan(size, auxiliary),
        "live-direct-lookup" => read::live_direct_lookup(size),
        "immutable-direct-lookup" => read::immutable_direct_lookup(size),
        "live-direct-scan" => read::live_direct_scan(size),
        "immutable-direct-scan" => read::immutable_direct_scan(size),
        "live-open" => read::live_open(size, auxiliary),
        "snapshot" => read::snapshot(size),
        "immutable-feed-random" => sdk::immutable_feed(size),
        "history-project" => sdk::history_project(size, auxiliary),
        "membership-matching-feeds" => sdk::matching_feeds(size, auxiliary),
        "membership-cardinalities" => sdk::aggregate_cardinalities(size, auxiliary),
        "membership-selected-pair" => sdk::aggregate_selected_pair(size),
        "membership-all-pairs" => sdk::aggregate_all_pairs(size, auxiliary),
        "direct-provider-join" => sdk::direct_join(size, auxiliary),
        "membership-provider-join" => sdk::membership_join(size, auxiliary),
        "algebra-count" => sdk::algebra_count(size, auxiliary),
        "algebra-compare" => sdk::algebra_compare(size, auxiliary),
        "algebra-publish-preserve" => sdk::algebra_publish(size, auxiliary, false),
        "algebra-publish-flat" => sdk::algebra_publish(size, auxiliary, true),
        "update-ipsets-workflow" => sdk::update_ipsets_workflow(size, auxiliary),
        _ => Err(format!("unknown scenario {name:?}")),
    }
}

pub(crate) fn reader_work(size: usize) -> Result<(usize, u64), String> {
    if size == 0 {
        return Err("reader benchmark size must be nonzero".to_owned());
    }
    let repetitions = 1_000_000usize.div_ceil(size);
    let units = size
        .checked_mul(repetitions)
        .ok_or_else(|| "reader benchmark work count overflows usize".to_owned())?;
    Ok((
        repetitions,
        u64::try_from(units).map_err(|_| "reader benchmark work count exceeds u64".to_owned())?,
    ))
}

pub(crate) fn count_points(
    size: usize,
    repetitions: usize,
    mut present: impl FnMut(Ipv4Key) -> Result<bool, String>,
) -> Result<u64, String> {
    let mut hits = 0u64;
    for _ in 0..repetitions {
        for index in 0..size {
            let address = Ipv4Key((index as u32).saturating_mul(4));
            hits += u64::from(present(address)?);
        }
    }
    Ok(black_box(hits))
}

pub(crate) fn count_cursor<C>(
    repetitions: usize,
    mut open: impl FnMut() -> Result<C, String>,
    mut advance: impl FnMut(&mut C) -> Result<bool, String>,
) -> Result<u64, String> {
    let mut records = 0u64;
    for _ in 0..repetitions {
        let mut cursor = open()?;
        while advance(&mut cursor)? {
            records += 1;
        }
    }
    Ok(black_box(records))
}

pub(crate) fn require_count(
    label: &str,
    observed: u64,
    expected: u64,
    noun: &str,
) -> Result<(), String> {
    if observed != expected {
        return Err(format!("{label} returned {observed} of {expected} {noun}"));
    }
    Ok(())
}

pub(crate) fn immutable_snapshot(
    database: &TestDatabase,
    records: usize,
) -> Result<PathBuf, String> {
    let output = database.snapshot();
    let snapshot = snapshot_to(
        database.main(),
        SnapshotSourceMode::Live,
        &output,
        SnapshotPublicationPolicy::FailIfExists,
        &snapshot_budget(records),
        &CancellationToken::new(),
    )
    .map_err(|failure| format!("{failure:?}"))?;
    if snapshot.cleanup_state() != iprange_livedb::publication::CleanupState::Clean {
        return Err(format!("snapshot cleanup is incomplete: {snapshot:?}"));
    }
    Ok(output)
}

pub(crate) fn result(
    name: &'static str,
    size: usize,
    auxiliary: usize,
    work_units: u64,
    database: &TestDatabase,
    measured: Measurement,
    file_path: &Path,
) -> Result<ScenarioResult, String> {
    validate_output(file_path, file_path == database.main())?;
    let (range_records, feeds) = live_info(database)?;
    let private_artifacts = database.private_artifacts()?;
    if private_artifacts != 0 {
        return Err(format!(
            "{name} left {private_artifacts} private temporary artifacts"
        ));
    }
    Ok(ScenarioResult {
        name,
        size,
        auxiliary,
        work_units,
        emitted_units: 0,
        range_records,
        feeds,
        measurement: measured,
        file: measure::file_size(file_path).map_err(|error| error.to_string())?,
        private_artifacts,
    })
}

pub(crate) fn immutable_result(
    spec: ImmutableResultSpec,
    database: &TestDatabase,
    measured: Measurement,
    file_path: &Path,
) -> Result<ScenarioResult, String> {
    validate_output(file_path, false)?;
    let reader =
        iprange_livedb::ImmutableReader::open(file_path).map_err(|error| error.to_string())?;
    let info = reader.info();
    drop(reader);
    let private_artifacts = database.private_artifacts()?;
    if private_artifacts != 0 {
        return Err(format!(
            "{} left {private_artifacts} private temporary artifacts",
            spec.name
        ));
    }
    Ok(ScenarioResult {
        name: spec.name,
        size: spec.size,
        auxiliary: spec.auxiliary,
        work_units: spec.work_units,
        emitted_units: spec.emitted_units,
        range_records: info.range_record_count,
        feeds: info.active_feed_count,
        measurement: measured,
        file: measure::file_size(file_path).map_err(|error| error.to_string())?,
        private_artifacts,
    })
}

fn validate_output(path: &Path, live: bool) -> Result<(), String> {
    let mode = if live {
        ValidationMode::LiveCurrent
    } else {
        ValidationMode::ImmutableCurrent
    };
    let mut sink =
        |_: &iprange_livedb::validation::ValidationFinding| Ok(ValidationSinkControl::Continue);
    let validated = validate(
        path,
        mode,
        &ValidationBudget::heap_only(64 * 1024 * 1024, 2),
        &iprange_livedb::CancellationToken::new(),
        &mut sink,
    )
    .map_err(|failure| format!("benchmark output validation failed: {failure:?}"))?;
    if !validated.valid {
        return Err(format!(
            "benchmark output has {} validation findings",
            validated.progress.finding_count
        ));
    }
    Ok(())
}

pub(crate) fn require_committed(commit: CommitResult) -> Result<(), String> {
    if commit.durability != CommitDurability::Committed {
        return Err(format!("commit did not complete: {commit:?}"));
    }
    Ok(())
}

pub(crate) fn close_writer(writer: &mut LiveWriter) -> Result<(), String> {
    let result = writer.close().map_err(|error| error.to_string())?;
    if result.outcome != CloseOutcome::Closed {
        return Err(format!("writer close did not complete: {result:?}"));
    }
    Ok(())
}

pub(crate) fn close_reader(reader: &mut LiveReader) -> Result<(), String> {
    let result = reader.close().map_err(|error| error.to_string())?;
    if result.outcome != CloseOutcome::Closed {
        return Err(format!("reader close did not complete: {result:?}"));
    }
    Ok(())
}

fn live_info(database: &TestDatabase) -> Result<(u64, u64), String> {
    let cancellation = iprange_livedb::CancellationToken::new();
    let mut reader =
        LiveReader::open(database.main(), &cancellation).map_err(|error| error.to_string())?;
    let info = reader.info().map_err(|error| error.to_string())?;
    close_reader(&mut reader)?;
    Ok((info.range_record_count, info.active_feed_count))
}

File: v4/go/cmd/iprange-v4-bench/scenarios.go
// Scenario registry and shared result plumbing (Rust
// benches/update_ipsets/scenarios.rs): every scenario registers itself
// under its Rust name, and the shared helpers reproduce the exact
// output validation, live-info counting, commit/close assertions, reader
// work shaping, and point/cursor counting of the Rust harness.
package main

import (
	"fmt"
	"sort"

	"github.com/firehol/iprange/v4/go"
)

type scenarioFunc func(size int, auxiliary int) (*scenarioResult, error)

var scenarioRegistry = map[string]scenarioFunc{}

func registerScenario(name string, fn scenarioFunc) {
	if _, exists := scenarioRegistry[name]; exists {
		panic("duplicate benchmark scenario " + name)
	}
	scenarioRegistry[name] = fn
}

type scenarioResult struct {
	Name             string
	Size             int
	Auxiliary        int
	WorkUnits        uint64
	EmittedUnits     uint64
	RangeRecords     uint64
	Feeds            uint64
	Measurement      measurement
	File             fileSize
	PrivateArtifacts uint64
}

type immutableResultSpec struct {
	Name         string
	Size         int
	Auxiliary    int
	WorkUnits    uint64
	EmittedUnits uint64
}

// result mirrors Rust scenarios::result: validate the output, count the
// live-range and feed facts, and refuse any leftover private artifacts.
func result(name string, size, auxiliary int, workUnits uint64, db *testDatabase, measured measurement, filePath string) (*scenarioResult, error) {
	if err := validateOutput(filePath, filePath == db.main); err != nil {
		return nil, err
	}
	rangeRecords, feeds, err := liveInfo(db)
	if err != nil {
		return nil, err
	}
	privateArtifacts, err := db.privateArtifacts()
	if err != nil {
		return nil, err
	}
	if privateArtifacts != 0 {
		return nil, fmt.Errorf("%s left %d private temporary artifacts", name, privateArtifacts)
	}
	file, err := fileSizeOf(filePath)
	if err != nil {
		return nil, err
	}
	return &scenarioResult{
		Name:             name,
		Size:             size,
		Auxiliary:        auxiliary,
		WorkUnits:        workUnits,
		RangeRecords:     rangeRecords,
		Feeds:            feeds,
		Measurement:      measured,
		File:             file,
		PrivateArtifacts: privateArtifacts,
	}, nil
}

// immutableResult mirrors Rust scenarios::immutable_result for snapshot
// outputs: validate, count facts through an immutable reader, refuse
// leftover private artifacts.
func immutableResult(spec immutableResultSpec, db *testDatabase, measured measurement, filePath string) (*scenarioResult, error) {
	if err := validateOutput(filePath, false); err != nil {
		return nil, err
	}
	reader, err := iprangedb.OpenImmutable(filePath)
	if err != nil {
		return nil, err
	}
	info, err := reader.Info()
	if err != nil {
		return nil, err
	}
	_ = reader.Close()
	privateArtifacts, err := db.privateArtifacts()
	if err != nil {
		return nil, err
	}
	if privateArtifacts != 0 {
		return nil, fmt.Errorf("%s left %d private temporary artifacts", spec.Name, privateArtifacts)
	}
	file, err := fileSizeOf(filePath)
	if err != nil {
		return nil, err
	}
	return &scenarioResult{
		Name:             spec.Name,
		Size:             spec.Size,
		Auxiliary:        spec.Auxiliary,
		WorkUnits:        spec.WorkUnits,
		EmittedUnits:     spec.EmittedUnits,
		RangeRecords:     info.RangeRecordCount,
		Feeds:            info.ActiveFeedCount,
		Measurement:      measured,
		File:             file,
		PrivateArtifacts: privateArtifacts,
	}, nil
}

// validateOutput mirrors Rust scenarios::validate_output: the output
// database must validate clean under the heap-only budget.
func validateOutput(path string, live bool) error {
	mode := iprangedb.ValidationModeImmutableCurrent
	if live {
		mode = iprangedb.ValidationModeLiveCurrent
	}
	var firstFinding string
	result, failure := iprangedb.Validate(path, mode, iprangedb.HeapOnly(64*1024*1024, 2), nil, iprangedb.SinkFunc(func(f *iprangedb.ValidationFinding) (iprangedb.ValidationSinkControl, error) {
		if firstFinding == "" {
			firstFinding = fmt.Sprintf("reason=%d object=%v", f.Reason, f.Object)
		}
		return iprangedb.SinkContinue, nil
	}))
	if failure != nil {
		return fmt.Errorf("benchmark output validation failed: %v", failure)
	}
	if !result.Valid {
		return fmt.Errorf("benchmark output has %d validation findings (first: %s)", result.Progress.FindingCount, firstFinding)
	}
	return nil
}

// requireCommitted mirrors Rust scenarios::require_committed.
func requireCommitted(commit iprangedb.CommitResult, err error) error {
	if err != nil {
		return err
	}
	if commit.Status != iprangedb.CommitCommitted {
		return fmt.Errorf("commit did not complete: %+v", commit)
	}
	return nil
}

// closeWriter mirrors Rust scenarios::close_writer.
func closeWriter(w *iprangedb.LiveWriter) error {
	result, err := w.Close()
	if err != nil {
		return err
	}
	if result.Outcome != iprangedb.CloseOutcomeClosed {
		return fmt.Errorf("writer close did not complete: %+v", result)
	}
	return nil
}

// closeLiveReader mirrors Rust scenarios::close_reader.
func closeLiveReader(r *iprangedb.LiveReader) error {
	result, err := r.Close()
	if err != nil {
		return err
	}
	if result.Outcome != iprangedb.CloseOutcomeClosed {
		return fmt.Errorf("reader close did not complete: %+v", result)
	}
	return nil
}

// liveInfo mirrors Rust scenarios::live_info: ranges and active feeds
// through a live reader on the main file.
func liveInfo(db *testDatabase) (uint64, uint64, error) {
	reader, err := iprangedb.OpenLiveReader(db.main, nil)
	if err != nil {
		return 0, 0, err
	}
	info, err := reader.Info()
	if err != nil {
		return 0, 0, err
	}
	if err := closeLiveReader(reader); err != nil {
		return 0, 0, err
	}
	return info.RangeRecordCount, info.ActiveFeedCount, nil
}

// readerWork mirrors Rust scenarios::reader_work.
func readerWork(size int) (int, uint64, error) {
	if size <= 0 {
		return 0, 0, fmt.Errorf("reader benchmark size must be nonzero")
	}
	repetitions := (1_000_000 + size - 1) / size
	return repetitions, uint64(size) * uint64(repetitions), nil
}

// countPoints mirrors Rust scenarios::count_points (dispersed sweep at
// index*4).
func countPoints(size, repetitions int, present func(iprangedb.IPv4) (bool, error)) (uint64, error) {
	var hits uint64
	for range repetitions {
		for index := 0; index < size; index++ {
			found, err := present(iprangedb.IPv4(uint32(index) * 4))
			if err != nil {
				return 0, err
			}
			if found {
				hits++
			}
		}
	}
	return hits, nil
}

// countRandomPoints mirrors Rust scenarios::count_random_points over the
// dispersed point list.
func countRandomPoints(points []uint32, repetitions int, present func(iprangedb.IPv4) (bool, error)) (uint64, error) {
	var hits uint64
	for range repetitions {
		for _, address := range points {
			found, err := present(iprangedb.IPv4(address))
			if err != nil {
				return 0, err
			}
			if found {
				hits++
			}
		}
	}
	return hits, nil
}

// countCursor mirrors Rust scenarios::count_cursor: one cursor per
// repetition, advancing until the end.
func countCursor[C any](repetitions int, open func() (*C, error), advance func(*C) (bool, error)) (uint64, error) {
	var records uint64
	for range repetitions {
		cursor, err := open()
		if err != nil {
			return 0, err
		}
		for {
			more, err := advance(cursor)
			if err != nil {
				return 0, err
			}
			if !more {
				break
			}
			records++
		}
	}
	return records, nil
}

// toPageBudget converts the bench page budget shape onto the public
// writer budget (Rust TransactionBudget parity).
func toPageBudget(b pageBudgetShape) iprangedb.PageBudget {
	return iprangedb.PageBudget{
		MaxHeapBytes:    b.MaxHeapBytes,
		MaxPrivatePages: b.MaxPrivatePages,
		MaxGrowthPages:  b.MaxGrowthPages,
		MaxOpenFiles:    b.MaxOpenFiles,
	}
}

// immutableSnapshot mirrors Rust scenarios::immutable_snapshot: live
// source to the canonical snapshot path, failing if it exists, clean
// cleanup required.
func immutableSnapshot(db *testDatabase, records int) (string, error) {
	output := db.snapshot()
	snap, err := iprangedb.SnapshotTo(db.main, iprangedb.SnapshotSourceLive, output, iprangedb.PolicyFailIfExists, &iprangedb.SnapshotBudget{
		MaxHeapBytes:   64 * 1024 * 1024,
		MaxOutputPages: snapshotBudget(records).MaxOutputPages,
		MaxOpenFiles:   3,
	}, nil)
	if err != nil {
		return "", err
	}
	if snap.CleanupState() != iprangedb.CleanupStateClean {
		return "", fmt.Errorf("snapshot cleanup is incomplete: %+v", snap)
	}
	return output, nil
}

// dispatchScenario runs the named scenario in-process (the child of the
// case/matrix drivers).
func dispatchScenario(name string, size, auxiliary int) (*scenarioResult, error) {
	if size <= 0 {
		return nil, fmt.Errorf("size must be positive")
	}
	fn, ok := scenarioRegistry[name]
	if !ok {
		return nil, fmt.Errorf("unknown scenario %q", name)
	}
	return fn(size, auxiliary)
}

func sortedScenarioNames() []string {
	names := make([]string, 0, len(scenarioRegistry))
	for name := range scenarioRegistry {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}
